### DISCUSSION TITLE
Collection IO

### DISCUSSION THREAD

**Discussion by narendasan** [pytorch/TensorRT#629] [RFCs]
# Collections

## Goal

Currently TRTorch programs that can be compiled must be trivially reducible to the form `f([Tensor]) -> [Tensor]`. Cases like `f(Tensor) -> ((Tensor, Tensor))` are supported through this method. This means that any sort of Input/Output formatting is not currently handled by TRTorch.  We would like to add support for cases like `f(Tensor[]) -> (Tensor, Tensor, (Tensor, Tensor)) ` or `f(Tensor, Tensor, (Tensor, Tensor)) -> (Tensor, (Tensor, Tensor)) ` which have non trivial subgrouping of tensors.

## API Considerations

Considering that the formatting of the function signature is now more complex, we might want to think about ways to make it easy to convey the input specification. 

### Proposed API

For a module with a signature such as:

```python
...
def forward(x, y, (a, b, c)):
	...
	
```

We could change the API to expect a tuple formatted in the same way someone might call the function. In conjunction with the example tensor feature (https://github.com/NVIDIA/TRTorch/pull/616), this might provide a natural way to reuse or more easily provide input specs vs. doing some sort of mental computation about aligning specs with inputs. 

#### Example

```python
x = trtorch.Input(<shape>)
y = trtorch.Input(<shape>)
a = torch.randn(<shape>)
b = torch.randn(<shape>)
c = torch.randn(<shape>)

trtorch.compile(mod, inputs=(x,y,(a,b,c)))
```

This is as opposed to 

```python
x = trtorch.Input(<shape>)
y = trtorch.Input(<shape>)
a = torch.randn(<shape>)
b = torch.randn(<shape>)
c = torch.randn(<shape>)

trtorch.compile(mod, inputs=[x,y,a,b,c])
```

Where the inputs must be aligned properly and paired internally with the graph signature.

The advantage is we can create an internal structure which encodes the format of the inputs for the user directly from the tuple provided. It also gives us an input of fixed size. Alternative methods that examine the graph input signature may have these fixed sizes obfuscated by type information. For instance the graph signature that uses a list to group  subsets of arguments instead of a tuple you might see a signature like:

```
graph(%x : Tensor, %y : Tensor, %abc : Tensor[]):
  ...
       %trt_ins : Tensor[] = prim::ListConstruct(%x, %y, %a, %b, %c)
       %trt_outs : Tensor[] = tensorrt::execute_engine(... %trt_ins)
 ...
  -> ((%i, (%j, %k))
```

This will not tell us how to align the inputs provided by the user as a flat list.

One limitation of this design may be the usage in C++, more exploration will be required to determine if this is ergonomic and consistent with PyTorch

## Internal Implementation

> Leverage TorchScript IR 

### 1.Inputs

We could look to make `trtorch::core::ir::Input` compatible with IValues by registering it as a torch custom class. This would let us nest Inputs in PyTorch types. This means we can pass around one IValue which holds the full input spec. This can then be parsed in the graph construction phase directly.

#### 1. Go from user spec to IValue

Its unclear the exact process to go from a presumably standard Python or C++ tuple to an IValue but this is something that PyTorch is able to do so it should just require looking at the source for PyTorch. 

#### 2. Assign IDs to Inputs and create list of Inputs to pass to TensorRT 

The next step is to populate a data structure like the one below which assigns each input an ID so that we can create a flattened vector of inputs to pass to TensorRT. 

We should add a field to the `trtorch::core::ir::Input` class which is called ID. This will be the unique identifier for the Input during compilation. The order in which we add these inputs will be determined by an in-order traversal of the tuple provided by the user. We only increment the id counter when we hit a new un-labeled Input (i.e. the leaves of the syntax tree). At the same time we can create a list of Inputs which will be passed to the conversion phase. This likely should be stored in a single struct

```c++
namespace trtorch {
namespace core {
namespace ir {

struct GraphInputs {
  torch::jit::IValue* input_signature;
  std::vector<Input> flattened_inputs;
};

} // namespace ir
} // namespace core
} // namespace trtorch
```

This object should then be added to the CompileSpec (this could potentially replace the vector of Inputs we use right now).

#### 2. Parse IValue and Construct Graph

Once we get to the graph construction phase we now need to amend it so that the first step is to create the input to graph and then take the inputs and flatten them to a list where each index of the list corresponds to the ID of each Input in TorchScript. This will involve using the IValue created in step 1 as the spec for access procedure for each Input. 

### 2. Outputs

```
%trt_out : Tensor[] = tensorrt::execute_engine(...) 
...
-> (x, y, [a, b], c)
```

> Returned from Conversion: `IValue`: torch::jit::IValue((0, 1 List[2, 3], 4))  

#### 1. Evaluating collection operations to get list of outputs

The evaluation system should automatically construct any sort of collections that will be used in the output during conversion. However currently MarkOutputs only handles ITensors and TensorContainers. It will need to extended to handle parsing the collection types. At this time we should construct a similar IValue to the Input IValue which encodes the indexes from the output of TensorRT to the final output tuple.  This IValue should be returned from the conversion process with the serialized TensorRT engine. We already have an ID for each output to deal with the fact that TensorRT doesn't guarantee output order. These IDs can be reused in the IValue. 

#### 2. Parse IValue and Construct Graph

In the graph construction phase once the TensorRT engine is embedded now we need to add the nodes to pack the outputs into the right format. This should use a similar system to the input system except it is packing Tensors from a list into a format vs unpacking. 

### Data Structures

```c++
namespace trtorch {
namespace core {
namespace ir {
  
struct GraphInputs {
  torch::jit::IValue input_signature;
  std::vector<Input> flattened_inputs;
};
  
typedef std::pair<GraphInputs, torch::jit::IValue> GraphIO;

} // namespace ir
} // namespace core
} // namespace trtorch
```

GraphIO is a pair where the first index is a struct which both the formatted input tuple containing `core::ir::Input` structs. And then a flattened version of the input tuple. The second index holds an IValue which is formatted tuple of Ints which defines how to go from the list output of TensorRT to the output tuple. 

### WAR 

We should first check to see if partial compilation can handle some of this trivially to start so that users can get unblocked

### MVP

We should implement support for one to two simple collection types. I think that tuples likely will be the simplest so we should start with that and get the system working end to end from user API to graph synthesis. 

### Additional Data Types

The next least complex type would be lists most likely. They should be implementable like tuples with very few changes if we use the API described above. After that we may want to look at dictionaries (this could be pushed to a later release even) which have the added complexity of keys. 

### Syntax Sugar 

Finally we should consider if there is any way to make the API simpler than what we have proposed here. If there is any work we could do for the user.

**Comment by narendasan** [comment]
cc: @inocsin @peri044

**Comment by inocsin** [comment]
Hi @narendasan, I have a few questions
(1) Do we need to change the input and output of tensorrt::execute_engine from vector<Tensor> to IValue?
(2) Do we change the code like this?
```c++
void AddEngineToGraph(
    torch::jit::script::Module mod,
    std::shared_ptr<torch::jit::Graph>& g,
    const std::string& serialized_engine,
    runtime::CudaDevice& device_info,
    GraphIO graph_io,
    std::string engine_id = "",
    bool fallback = false)

{
...
// Add inputs to graph
// Setting the input binding relation
torch::jit::Value nested_inputs = xxx;
g->block()->appendNode(nested_inputs);
...
  auto execute_node = g->create(
      c10::Symbol::fromQualString("tensorrt::execute_engine"),
      torch::jit::ArrayRef<torch::jit::Value*>(execute_node_inputs),
      1);
...
// Set the output binding relation
// Register outputs
...
}

```

**Comment by chaoz-dev** [comment]
A couple of questions regarding this spec for input: 

1. Do we _need_ to support mixing tensors and containers of tensors on the input? I made a similar suggestion in #798, but thinking about this some more, perhaps we can shortcut this implementation by just supporting uniform input types? ie. 
```
def forward( input: Union[ *torch.Tensor, List[torch.Tensor], Tuple[torch.Tensor], namedtuple[torch.Tensor]] )
```
This may significantly simplify the implementation (for MVP at least), as we should need only to unwrap the input container to reduce the problem to a form that we can already handle today (ordering is kept as given in the input container). I suspect that this may solve most common use cases, since the user may just append to their container should they need to mix tensors with tensor containers. 

The input shapes also have a natural 1-1 mapping to the inputs here as well.

2. Related to (1), is it overly naive to think that we could perform all of the container unwrapping prior to jit scripting on input? On the Python side, this seems fairly doable; I haven't looked at the C++ definition yet. But if so, then we could reduce the containers into raw tensor inputs, in which case we could side-step the need to change any internal types as the input going into compilation will just look like they do now (ie. multiple separate `torch.Tensor`s).

**Reply by narendasan** [reply]
I think long term we definitely want to support mixing collections. I can think of cases like LSTMs where this might be natural. Not sure how many cases get unblocked if you just add a switch for different variants of collections and just assume one collection of inputs, but it might be a good start until @inocsin's work matures. Overall we need to do the work in the core library so that both Python and C++ APIs will work properly as well as we need to construct the same interface in the compiled module

**Comment by chaoz-dev** [comment]
@inocsin How far along on the implementation of this are you already? Would you happen to have a public dev branch that is usable? If not, I don't mind taking a crack at the implementation here, focusing specifically on input for now.

**Reply by inocsin** [reply]
It is still in progress, here is the pr https://github.com/NVIDIA/Torch-TensorRT/pull/802

**Reply by chaoz-dev** [reply]
Ah perfect, thanks!

**Comment by inocsin** [comment]
For a simple model with tuple as input as below
```python
import torch
import copy
import torch.nn as nn
import torch.nn.functional as F
from typing import Tuple, List, Dict

class TestModel(nn.Module):
    def __init__(self):
        super(TestModel, self).__init__()

    def forward(self, z: Tuple[torch.Tensor, torch.Tensor]):
        r = z[0] + z[1]
        return r

test_model = TestModel()

ts = torch.jit.script(test_model)
print(ts.graph)
```
The original graph is
```
graph(%self : __torch__.___torch_mangle_16.TestModel,
      %z.1 : (Tensor, Tensor)):
  %3 : int = prim::Constant[value=0]() # /tmp/ipykernel_37045/2552691859.py:13:14
  %6 : int = prim::Constant[value=1]() # /tmp/ipykernel_37045/2552691859.py:13:21
  %4 : Tensor = prim::TupleIndex(%z.1, %3)
  %7 : Tensor = prim::TupleIndex(%z.1, %6)
  %r.1 : Tensor = aten::add(%4, %7, %6) # /tmp/ipykernel_37045/2552691859.py:13:12
  return (%r.1)
```
If we use `torch::jit::LowerAllTuples`[link](https://github.com/NVIDIA/Torch-TensorRT/blob/master/core/lowering/lowering.cpp#L35), we will get the lowered graph
```
DEBUG: [Torch-TensorRT TorchScript Conversion Context] - graph(%7 : Tensor,
      %8 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %6 : Tensor = aten::add(%7, %8, %2)
  return (%6)
```

And models that take list as input as below will produce `aten::__getitem__`. 
```python
class TestModel(nn.Module):
    def __init__(self):
        super(TestModel, self).__init__()

    def forward(self, z: List[torch.Tensor]):
        r = z[0] + z[1]
        return r
```
```
graph(%self : __torch__.___torch_mangle_18.TestModel,
      %z.1 : Tensor[]):
  %3 : int = prim::Constant[value=0]()
  %6 : int = prim::Constant[value=1]()
  %4 : Tensor = aten::__getitem__(%z.1, %3)
  %7 : Tensor = aten::__getitem__(%z.1, %6)
  %r.1 : Tensor = aten::add(%4, %7, %6)
  return (%r.1)
```

When I was implementing this feature, I have a few questions as below, do you have any suggestions? Thanks. @narendasan 
(1) In order to mapping tuple/list to a flatten one, do we need to evaluate the value of index in `aten::__getitem__` and `prim::TupleIndex` to find the relationship between Input spec and items in tuple?
(2) In MapInputsAndDetermineDTypes[https://github.com/inocsin/TRTorch/blob/collection/core/compiler.cpp#L414] we need to set input data type. When the input is tuple, should we just set the data type of Value like `%4 : Tensor = prim::TupleIndex(%z.1, %3)`, and ignore the orignal tuple

**Reply by narendasan** [reply]
What is the assumed format of the input specs? I would think for 1. yes you would basically evaluate those ops to determine the mapping. 2. Yes that would be ideal. As long as you can still maintain enough information to reconstruct the correct unpacking and repacking of the input tensors

**Comment by chaoz-dev** [comment]
Hey wanted to follow up on this since it's been a while; have there been any additional updates to this?

**Reply by narendasan** [reply]
@inocsin do you have an update to share?

**Reply by inocsin** [reply]
Still need some time. I'm working on binding the elements of tuple and their ir::Input.
I have changed the GraphInputs struct, elements in inputs like `((a,b,c), [d, e, f])` can be more easily indexed without using offset.

```
struct GraphInputs {
  torch::jit::IValue* input_signature; // like ((a,b,c), [d, e, f])
  std::vector<Input> flattened_inputs;  // old
  std::vector<std::vector<ir::Input>> collection_inputs; // new
};
```

**Comment by narendasan** [comment]
As of v1.2.0 there will be experimental collections support (#1201). It comes with the following caveats:

> Partial compilation must be enabled and the following operations will be forced to run in PyTorch:
```    
    - aten::__getitem__
    - prim::ListConstruct
    - prim::ListUnpack
    - prim::TupleIndex
    - prim::TupleConstruct
    - prim::TupleUnpack
```

This means users will not have access to features like dynamic shape and performance may not be optimal. The intention is to address this limitation in v1.3.0

**Comment by narendasan** [comment]
When we go to implement the graph synthesis component we should include the ability to handle optional tensors which may or may not be `None` (#772). For example given a forward function that uses optionals: 
```py
def forward(self, x: Tuple[torch.Tensor, torch.Tensor], y: Optional[torch.Tensor], z: List[Optional[torch.Tensor]]):
```
A user should be able to do this 
```py
torch_tensorrt.compile(mod,
    input_signature=((a,b), None, [c, None, d],)
)
```

**Reply by narendasan** [reply]
The `GraphInputs` object should be able to create a mapping from each of the 6 inputs to an index in the TensorRT input list and for `None`'s should insert a `None` into the `evaluated_value_map`

**Comment by gs-olive** [comment]
## Further Work and Suggestions
In regard to models which output Tuples or other complex types, certain unexpected failures can stem from specifying `require_full_compilation=True`. Specifically, in the case of the [HuggingFace T5-base](https://huggingface.co/t5-base) model, as discussed in Issue #1583, when full support for its operators is implemented, as in PR #1584, using the `require_full_compilation=True` flag causes the model compilation to fail, with the following message:
```python
DEBUG: [Torch-TensorRT TorchScript Conversion Context] - Evaluating %729 : (Float(1, 128, 768, strides=[98304, 768, 1], requires_grad=1, device=cpu)) = prim::TupleConstruct(%input)
DEBUG: [Torch-TensorRT TorchScript Conversion Context] - Found the value to be: (<__torch__.torch.classes._torch_tensorrt_eval_ivalue_types.TensorContainer object at 0x881822b0>,)
RuntimeError: [Error thrown at core/conversion/conversion.cpp:230] Tuple type. Only a single tensor or a TensorList type is supported.
```
This error does not appear, however, when `require_full_compilation=False`, as the Tuple output construction defaults to run in Torch.

To resolve this issue, it could be helpful to modify the `MarkOutputs` function in conversion: 
https://github.com/pytorch/TensorRT/blob/5063b145ad1935f440a2fca1b4ed0a92ba9e7139/core/conversion/conversion.cpp#L214-L246
If this function can handle tuple-formatted outputs, and more generally, if Input/Output tensor formatting could be handled symbolically, as is discussed in the RFC and comments above, this could be a promising approach to resolve the above error as well as other related ones. Additionally, this would provide a performance boost to such models, as requiring a Torch-executed block simply for casting inputs and outputs to the correct nesting format could slow down inference. 

Depending on the scope of the solution, this could require updates to `core/compiler` and the `ir` Input class to handle abstract syntactical representations of inputs and outputs, as well as a method of translating back and forth between flattened tensor lists and their nested counterparts.

**Reply by gs-olive** [reply]
### Cases
There are multiple cases to be aware of and address when allowing the flag `require_full_compilation=True` with complex IO collection types.  We currently have evaluators for multiple collection-based operators, such as `prim::TupleUnpack`, however these are disabled in favor of executing collections operators in Torch. Specifically, since TensorRT does not support collection-type inputs, we require a mapping from the IO signature for each TRT-executed block, to the raw input Tensors passed to TensorRT. There are a few cases to note, as detailed below:
#### 1. Collection Input [i.e. (Tensor, (Tensor, Tensor)) $\to$ Tensor]
This format is experimentally supported as of release 1.2, via the `input_signature` argument. It is experimental, and currently the Input `(Tensor, (Tensor, Tensor))` is not functional.

#### 2. Collection Output [i.e. Tensor, Tensor $\to$ (Tensor, (Tensor, Tensor))]
This is not supported with `require_full_compilation=True`, due to an error in marking the output.

#### 3. Collection IO + Internal Tensors [i.e. (Tensor, (Tensor, Tensor)) $\to$ (Tensor, (Tensor, Tensor))]
This is not supported with `require_full_compilation=True`, as both of the above would need to be fully supported first.

The benefit to supporting Collection IO, particularly for internal tensors, is that TRT-operator-supported graphs would not have to be segmented into suboptimal blocks due to collection Packing/Unpacking, potentially costing a great deal of performance/inference time.

**Comment by gs-olive** [comment]
## Dictionary `{str: Tensor}` Input Support in TorchScript
### Context
Currently, we do not support inputs which are of the form `Dict[str, Tensor]`. The challenge with these inputs is primarily the presence of the strings, as we cannot currently provide string inputs to the backend `core`/`ir`.
### Necessary Work for Full Support [Small/Medium]
One approach to supporting dictionary inputs is to refactor `input_signature`. Specifically - as opposed to reconstructing the input (as `_parse_input_signature` does currently), we would preferably parse the input structure as an `IValue` and store a data structure which indicates the shape/dtype of each tensor input. Alternatively, for the Python API, we could add logic to manually refactor dictionary inputs into flattened lists of Tensor inputs which are then reconstructed into dictionaries via wrapper modules (see section below).
### Effective Workaround
Currently, an effective workaround exists which avoids the need to fully support dictionary inputs (and most complex collection inputs). Since the collection packing/unpacking operations can take place in Torch, we can reformat the input in a manner which allows Torch-TRT to effectively parse it, by writing a function wrapper:
```python
class DictionaryInputModel(torch.nn.Module):
    def __init__(self):
        super(DictionaryInputModel, self).__init__()

    def forward(self, d: Dict[str, Tensor]):
        a = d["x"] +  d["x"]
        a +=  d["y"]
        return a
```
Compiling the above in Torch-TRT will fail, as `input_signature` does not parse dictionaries. However, we can wrap the model as such:
```python
class WrappedModel(torch.nn.Module):
    def __init__(self):
        super(WrappedModel, self).__init__()
        self.wrapper = DictionaryInputModel()

    def forward(self, x: Tensor, y: Tensor):
        dictionary = {"x": x, "y": y}
        return self.wrapper(dictionary)
```
Now, compiling an instance of `WrappedModel()` is straightforward and functional, as the inputs have been flattened to a list. The dictionary-building operations are handled within Torch blocks, which are delegated during partitioning, thus providing an alternative to full dictionary input support.
